engine/src/test/java/com/datatorrent/stram/StramLocalClusterTest.java

```diff
@@ -381,8 +382,11 @@ private String generatejar(String pojoClassName) throws IOException, Interrupted
     String sourceDir = "src/test/resources/dynamicJar/";
     String destDir = testMeta.getPath();
 
+    // The compiled java class should be loadable by the current java runtime hence setting the compile target version
+    // to be the same
+    String version = getJavaSrcVersion();
     Process p = Runtime.getRuntime()
-        .exec(new String[] {"javac", "-d", destDir, sourceDir + pojoClassName + ".java"}, null, null);
+        .exec(new String[] {"javac", "-source", version, "-target", version, "-d", destDir, sourceDir + pojoClassName + ".java"}, null, null);
```

Comment:
It only works if "javac" on the PATH is the same or later than the version used to run unit tests. It is not always the case and version of javac on the PATH may be 1.6 or 1.7. Instead of using external "javac" can it rely on `JavaCompiler`?

Reply:
Did think about programmatic invocation of the java compiler as well but reasoned that since this is a test case and it is a quite unusual case that javac would be a different version than the underlying java runtime being used to run the tests, this workaround would suffice.

Reply:
Please add an explicit path to "javac" and "jar" instead of relying on one to be on the PATH. Is it necessary to specify "source" version?

Reply:
There isn't a direct way to find out an appropriate path of javac, you have to make some assumptions which may not be correct, for example, JAVA_HOME may not be defined to look in $JAVA_HOME/bin. It may or may not reside in the same folder as the java executable, if a separate jre is being used for the runtime and also path of java itself may not be directly determinable without looking into process information. For these reasons I think you would have to rely on javac being in the PATH which is what the test is doing.

I tested openjdk and it needs the source version specified as well when the target is a lower version.

Reply:
The assumption that the right version of "javac" is on the PATH is more likely to be incorrect (I've seen such configuration on Apache Jenkins). It is not necessary to use `${JAVA_HOME}` environment variable that may not be defined as you already mentioned. Instead, consider using "java.home" property.

Reply:
I am not sure it is safe to rely on javac being in prop{java.home}/bin. If a jre is being used there will be no javac there.

Reply:
Also the pom.xml settings will ensure that the javac in the path at least supports 1.8.

Reply:
IMO, it is equally unreliable to assume that "javac" is on the PATH and PATH is properly setup to point to the right version of "javac" or "jar". It is more likely that "java" used to execute the test is from the same installation that was used to compile it either in maven or IDE.

Reply:
AFAIK, maven uses JAVA_HOME if defined.